### DISCUSSION TITLE
How to

### DISCUSSION THREAD

**Discussion by pink-fr3ud** [RudolphRiedel/FT800-FT813#36] [Q&A]
Hi!
Beginner here, trying to find some middle ground help between the BT81X programming guide and Rudolph library. The BRTcommunity forum being double locked (5day to pass moderator approval for each message), here I am

Setup: ESP32 using Arduino framework, Riverdi 7" RVT70HSBFWCA0

Rudolph demo works fine, now trying to do my own stuff by modifying his example. I'm trying to do a car dashboard. 
I made all my gauges background and needles images. They are all converted using EAB to COMPRESSED_RGBA_ASTC_8x8_KHR, I made a flash image and pushed it to the screen. I'm able to display them from flash no problem until I try to display more than 3 (304x304px image)

How could you help me? 
It's not really clear to me how all things work and the strategy to adopt:

-When should I use the "normal" display list or the burst one? How to know if my list is too long?
-When should I use the DL_END ?
-After some time the program crashes and only buttons are visible
-Why more than 3 images from flash make a repetitive pattern of the previous images? Should I flush something between calls of setbitmaps?
-Why do I have to do x16 on coordinates of bitmaps ? -> I'm already using VERTEX_FORMAT(0)

In a second phase:
-Should I make a static background for each page with gauges bitmaps ? Can I make it using Flash?
-Should I make a dynamic portion TFT_display where the needles move according to value, using maybe GRAM this time for faster compute time?

In a third phase:
-Swipe between pages (I use buttons for the moment), but the scrollbar doesn't "lock" into pages

images:
https://ibb.co/WpzBk0q  : 6 bitmaps problem
https://ibb.co/McYKkmp : 3 small bitmaps
https://ibb.co/YD6pvsc : 1 large bitmaps

```
#include "EVE.h"

/* some pre-definded colors */
#define RED 0xff0000UL
#define ORANGE 0xffa500UL
#define GREEN 0x00ff00UL
#define BLUE 0x0000ffUL
#define BLUE_1 0x5dade2L
#define YELLOW 0xffff00UL
#define PINK 0xff00ffUL
#define PURPLE 0x800080UL
#define WHITE 0xffffffUL
#define BLACK 0x000000UL

#define PRECISION 0

// Flash MAP
#if 1
typedef struct {
 uint8_t index;
 uint8_t tag;
 uint32_t addressFlash;
 uint32_t addressRamg;
 uint16_t w;
 uint16_t h;
 uint16_t format;
 } flash_Obj;
 
#define FLASH_ITEMS 46

flash_Obj* all_OBJ[FLASH_ITEMS];		//	id	tag?	Flash	RAMG	w	h			format							
	flash_Obj BLOB 					 	 =	{0,		0,		0, 		0,	 0,	 	  0,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	0	:	4096
	flash_Obj AFR_304					 =	{1,		0,	  4096,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	4096	:	23104
	flash_Obj AFR_336					 =	{2,		0,	 27200,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	27200	:	28224
	flash_Obj AFR_512					 =	{3,		0,	 55424,		0,	512,	512,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	55424	:	65536
	flash_Obj AFRarc_304				 =	{4,		0,	120960,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	120960	:	23104
	flash_Obj AFRarc_336				 =	{5,		0,	144064,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	144064	:	28224
	flash_Obj AFRarc_512				 =	{6,		0,	172288,		0,	512,	512,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	172288	:	65536
	flash_Obj BAT_304					 =	{7,		0,	237824,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	237824	:	23104
	flash_Obj BAT_336					 =	{8,		0,	260928,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	260928	:	28224
	flash_Obj BATarc_304				 =	{9,		0,	289152,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	289152	:	23104
	flash_Obj BATarc_336				 =	{10,	0,	312256,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	312256	:	28224
	flash_Obj BSTB_304					 =	{11,	0,	340480,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	340480	:	23104
	flash_Obj BSTB_336					 =	{12,	0,	363584,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	363584	:	28224
	flash_Obj BSTB_512					 =	{13,	0,	391808,		0,	512,	512,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	391808	:	65536
	flash_Obj BSTP_304					 =	{14,	0,	457344,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	457344	:	23104
	flash_Obj BSTP_336					 =	{15,	0,	480448,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	480448	:	28224
	flash_Obj BSTP_512					 =	{16,	0,	508672,		0,	512,	512,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	508672	:	65536
	flash_Obj BSTParc_304				 =	{17,	0,	574208,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	574208	:	23104
	flash_Obj BSTParc_336				 =	{18,	0,	597312,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	597312	:	28224
	flash_Obj BSTParc_512				 =	{19,	0,	625536,		0,	512,	512,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	625536	:	65536
	flash_Obj Image_1024				 =	{20,	0,	691072,		0,	1024,	464,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	691072	:	118784
	flash_Obj GTMP_304					 =	{21,	0,	809856,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	809856	:	23104
	flash_Obj GTMP_336					 =	{22,	0,	832960,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	832960	:	28224
	flash_Obj GTMParc_304				 =	{23,	0,	861184,		0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	861184	:	23104
	flash_Obj GTMParc_336				 =	{24,	0,	884288,		0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	884288	:	28224
	flash_Obj NDL_BAT_AFR_OILP_304_144	 =	{25,	0,	912512,		0,	144,	144,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	912512	:	5184
	flash_Obj NDL_BAT_AFR_OILP_336_160	 =	{26,	0,	917696,		0,	160,	160,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	917696	:	6400
	flash_Obj NDL_BAT_AFR_OILP_512_248	 =	{27,	0,	924096,		0,	248,	248,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	924096	:	15424
	flash_Obj NDL_BST_304_208			 =	{28,	0,	939520,		0,	208,	208,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	939520	:	10816
	flash_Obj NDL_BST_336_232			 =	{29,	0,	950336,		0,	232,	232,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	950336	:	13504
	flash_Obj NDL_BST_512_352			 =	{30,	0,	963840,		0,	352,	352,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	963840	:	30976
	flash_Obj NDL_WTMP_OTMP_GTMP_304_208 =	{31,	0,	994816,		0,	208,	208,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	994816	:	10816
	flash_Obj NDL_WTMP_OTMP_GTMP_336_232 =	{32,	0,	1005632,	0,	232,	232,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1005632	:	13504
	flash_Obj OILP_304					 =	{33,	0,	1019136,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1019136	:	23104
	flash_Obj OILP_336					 =	{34,	0,	1042240,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1042240	:	28224
	flash_Obj OILParc_304				 =	{35,	0,	1070464,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1070464	:	23104
	flash_Obj OILParc_336				 =	{36,	0,	1093568,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1093568	:	28224
	flash_Obj OTMP_304					 =	{37,	0,	1121792,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1121792	:	23104
	flash_Obj OTMP_336					 =	{38,	0,	1144896,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1144896	:	28224
	flash_Obj OTMParc_304				 =	{39,	0,	1173120,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1173120	:	23104
	flash_Obj OTMParc_336			 	 =	{40,	0,	1196224,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1196224	:	28224
	flash_Obj PeppaPIG_1024				 =	{41,	0,	1224448,	0,	1024,	576,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1224448	:	147456
	flash_Obj WTMP_304					 =	{42,	0,	1371904,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1371904	:	23104
	flash_Obj WTMP_336					 =	{43,	0,	1395008,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1395008	:	28224
	flash_Obj WTMParc_304				 =	{44,	0,	1423232,	0,	304,	304,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1423232	:	23104
	flash_Obj WTMParc_336   			 =	{45,	0,	1446336,	0,	336,	336,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1446336	:	28224
	flash_Obj PeppaPIG_VID				 =	{46,	0,	1474560,	0,	1024,	600,	EVE_COMPRESSED_RGBA_ASTC_8x8_KHR};	//	:	1474560	:	10530048
#endif

uint16_t num_profile_a, num_profile_b;
uint8_t tft_active = 0;
uint8_t page = 0;
uint8_t previous_page = 1;

void DrawObj(flash_Obj gauge,uint16_t x, uint16_t y)
{
	EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
	EVE_cmd_setbitmap_burst(0x800000 | (gauge.addressFlash / 32), gauge.format, gauge.w, gauge.h);
	EVE_cmd_dl_burst(VERTEX2F(x*16, y*16));
	EVE_cmd_dl_burst(DL_END);
}

// screen initialization
void TFT_init(void)
{
	if (EVE_init() != 0)
	{
		tft_active = 1;
		EVE_memWrite8(REG_PWM_DUTY, 0x30); /* setup backlight, range is from 0 = off to 0x80 = max */
		EVE_init_flash();
	}
}

// check for touch events and setup vars for TFT_display() 
void TFT_touch(void)
{
	uint8_t tag;
	static uint8_t toggle_lock = 0;

	if (EVE_busy()) /* is EVE still processing the last display list? */
	{
		return;
	}
	tag = EVE_memRead8(REG_TOUCH_TAG); /* read the value for the first touch point */

	switch (tag)
	{
	case 0:
		toggle_lock = 0;
		break;

	case 10:
		if (toggle_lock == 0)
		{
			toggle_lock = 42;

			if (page > 0)
			{
				page--;
			}
		}
		break;

	case 20:
		if (toggle_lock == 0)
		{
			toggle_lock = 42;
			if (page < 5)
			{
				page++;
			}
		}
		break;
	}
}

//	dynamic portion of display-handling, meant to be called every 20ms or more
void TFT_display(void)
{	
	if (tft_active != 0)
	{
		if (page != previous_page)
		{
		previous_page = page;
		EVE_start_cmd_burst(); // start writing to the cmd-fifo as one stream of bytes, only sending the address once 
		EVE_cmd_dl_burst(CMD_DLSTART); // start the display list 
		EVE_cmd_dl_burst(TAG(0));
        EVE_cmd_dl_burst(DL_CLEAR_RGB | BLACK); // set the default clear color to Black 
        EVE_cmd_dl_burst(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG); // clear the screen - this and the previous prevent artifacts between lists, Attributes are the color, stencil and tag buffers 
        EVE_cmd_bgcolor(BLACK); 
		EVE_cmd_dl(VERTEX_FORMAT(PRECISION)); // reduce precision for VERTEX2F to 1 pixel instead of 1/16 pixel default 
	
		switch (page)
		{
		case 0: // MAIN PAGE - 3 GAUGES //
			DrawObj(BAT_336,0,0);
			DrawObj(WTMP_336,344,0);
			DrawObj(OILP_336,687,0);
			break;

		case 1:	// 6 GAUGES  //
			DrawObj(OTMP_304,0,0);
			DrawObj(WTMP_304,360,0);
			DrawObj(OILP_304,719,0);
			
			DrawObj(BSTP_304,0,305);
			DrawObj(GTMP_304,360,305);
			DrawObj(AFR_304,720,305);
			break;

		case 2: // 1 LARGE GAUGE //
			DrawObj(BSTP_512,256,0);
			break;

		case 3: // LARGE IMAGE //
			DrawObj(Image_1024,0,0);
			break;
		
		case 4: 
			break;

		case 5: 
			break;	
		}
		
		EVE_cmd_dl_burst(TAG(10));
		EVE_cmd_button_burst(0, 540, 100, 60, 24, 0, "Previous");
		EVE_cmd_dl_burst(TAG(20));
		EVE_cmd_button_burst(924, 540, 100, 60, 24, 0, "Next");
	
		EVE_cmd_dl_burst(DL_DISPLAY); // instruct the graphics processor to show the list 
		EVE_cmd_dl_burst(CMD_SWAP);	  // make this list active 
		EVE_end_cmd_burst();		  // stop writing to the cmd-fifo, the cmd-FIFO will be executed automatically after this or when DMA is done 
		}
	}
	
}

```

Thank you !

**Comment by RudolphRiedel** [comment]
>The BRTcommunity forum being double locked (5day to pass moderator approval for each message),

Well, it was the holiday season.
But yes, a little less moderation would be nice.

>Setup: ESP32 using Arduino framework, Riverdi 7" RVT70HSBFWCA0

And annother one, I might need to dig out my ESP32 board from my hardware pile. :-)

>-When should I use the "normal" display list or the burst one?

Do you really mean display list or more broadly commands?
This is intended for the set of commands that form a display list which is also why
not every function comes with a ..._burst() variant.

The difference is using DMA to transfer the list or not.
You have DMA support available so you can use it.

While it looks like I am building two dispay lists in my example, only the one in TFT_display()
is using the ..._burst() functions while the commands in initStaticBackground() do not.
The reason ist that TFT_display() is for the refresh and therefore called over and over again so
having it use DMA really makes a difference.
The one in initStaticBackground() on the other hand is only a fragment of a display list that is only generated once.
And this fragment is copied in TFT_display() by
 EVE_cmd_append_burst(MEM_DL_STATIC, num_dl_static);
at the beginning of the real display list.
This is just one way to reduce the SPI traffic by avoiding sending parts of the display list over and over again that
are not changing anyways.

>How to know if my list is too long?

My simple demo prints a value for "Bytes" and a value for "DL-size" as a debugging measure.
The value for "Bytes" must be kept under 4080. And the number for "DL-size" must be kept unter 8192.

>-When should I use the DL_END

Whenever you are done with one group of objects that you started with
EVE_cmd_dl_burst(DL_BEGIN | EVE_...);
So can group several bitmaps together or lines or points or rectangles and so on.
But you should not start using bitmaps and then start using points with no DL_END in between.

>-After some time the program crashes and only buttons are visible

Ok, this should not happen.

>-Why more than 3 images from flash make a repetitive pattern of the previous images?

There appears to be a bandwidth issue with the externally attached flash, I am afraid that this
just does not work.

>Should I flush something between calls of setbitmaps?

Copy the images to RAM_G before you use them, at least there are rather small as ASTC and you might not even
need to copy all of them.

>-Why do I have to do x16 on coordinates of bitmaps ? -> I'm already using VERTEX_FORMAT(0)

Because you are using 
 EVE_cmd_dl(VERTEX_FORMAT(PRECISION));
after EVE_start_cmd_burst() which means that EVE_cmd_dl() ends up doing nothing at all.

>-Should I make a static background for each page with gauges bitmaps?

This could help reduce the SPI traffic but it may not be worth the effort,
you have to check whether there is an issue before starting to optimize.

>- Can I make it using Flash?

Unfortunately still now.

>-Swipe between pages (I use buttons for the moment), but the scrollbar doesn't "lock" into pages

The scrollbar locks wherever and however you want it to lock.
You need to use CMD_TRACK to well, track the movement of a finger when touching an object if a touch
is detected on that object.
While tracking is active you get a 16 bit value for the movement of the finger in REG_TRACKER as well as the tag-value.
You use this value to calculate  what position you are at and also the new "val" value for your scollbar.

Try this in the EVE Screen Editor:
CLEAR(1, 1, 1)
TAG(10)
CMD_SCROLLBAR(100, 100, 400, 16, 0, 100, 200, 500)
CMD_TRACK(100,100, 400, 20, 10)

Switch to touch in the toolbar at the top and change to the "Inspector" tab at the bottom.
Scroll down in the "RAM_REG" window to see the value for "REG_TRACKER".
When you use the mouse to generate touch events you will see that you only get values in REG_TRACKER when you are "touching" the scrollbar.
And the values are from 0 to 65535.
Now the scrollbar is set for 5 positions, a width of 100, 500 in total and sitting at 200 which is position 2.
Divide the tracking value from REG_TRACKER by 13107 as 65535 / 5 = 13107.
Hmm, ok, 65535 / 5 = 13107,000, so a tracking value of 65535 would actually result in a value of 5.
So either divide by 13108 or subtract one from the tracking value before dividing.
So we get a value of zero to four out of this.

This means the actual command in my display-list would look like this:
EVE_cmd_scrollbar_burst(100, 100, 400, 16, 0, scrollbar_position * 100, 200, 500);

And no, I am not checking if all the numbers work out completely nicely. :-)

Since CMD_TRACK would be keeping EVE busy for no reason when it would be active all the time and since one might want
to use different scrollbars or sliders, I am only activating the tracking in TFT_touch() when a touch of an object is detected.

Like this:

[code]
case 0:
	toggle_lock = 0;
	if(track != 0 )
	{
		EVE_cmd_track(0, 0, 0, 0, 0); // stop tracking
		track = 0;
	}
	break;

case 40:
	if(track == 0)
	{
		EVE_cmd_track(730,240,60,200,40); /* start tracking */
		track = 10;
	}
	else
	{
		uint32_t tracker_val = EVE_memRead32(REG_TRACKER);
		if((tracker_val & 0xff) == 40)
		{
			tracker_val = tracker_val / 1073725; /* -> 4000 max */
		}
	}
[/code]

Well, not the best example, just one example. the tracker value is used as 32 bit and divided as such instead of isolating
the upper 16 bit first and dividing afterwards.

**Reply by pink-fr3ud** [reply]
What a progress i made thank of your help !

- DMA, I wasn't using it, and indeed it's crazy
- Understood for _burst and list ;)
- I mimicked the initStaticBackground() you did by loading gauges faces in it after each page change, and it's a lot better! 

> My simple demo prints a value for "Bytes" and a value for "DL-size" as a debugging measure.
The value for "Bytes" must be kept under 4080. And the number for "DL-size" must be kept unter 8192.

- Bytes are reading 180 (doing nothing yet) and DL-Size is 1248 for 3 gauges displayed so pretty good!

>-When should I use the DL_END

- I didn't managed to group images, doesn't seems to work like this :
```EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
 EVE_cmd_setbitmap(0x800000 | (gauge0.addressFlash / 32), gauge0.format, gauge0.w, gauge0.h);
 EVE_cmd_dl(VERTEX2F(x0, y0));
 EVE_cmd_setbitmap(0x800000 | (gauge1.addressFlash / 32), gauge1.format, gauge1.w, gauge1.h);
 EVE_cmd_dl(VERTEX2F(x1, y1));
 EVE_cmd_dl(DL_END);
```
Is it only for the _burst list building?

> Copy the images to RAM_G before you use them, at least there are rather small as ASTC and you might not even need to copy all of them.

- I tried to load the 6 images from flash and put them in the MEM_DL_STATIC but no chance again, could it be because part of the picture are out of screen bonds? (1024*600 display, pictures are 304*304)

> Because you are using
 EVE_cmd_dl(VERTEX_FORMAT(PRECISION));
after EVE_start_cmd_burst() which means that EVE_cmd_dl() ends up doing nothing at all. 

- Good caught !

>After some time the program crashes and only buttons are visible

Still happens but it takes more time now. Bitmaps disappears but the rest of the screen still works ok

> Scrollbar

- I started looking at ESE, it's great tool to look at things fast! After playing a bit with it, I don't think the scrollbar is the feel I'm after, but I saw that there was a XY tracking coordinate. I'm going to try to make a Page+/ Page- by looking at the start XY touch and comparing it to the stop XY to get the direction 

Thanks a lot for your help Rudolph, i'm sure I will need it again If I don't find anything on the interweb !

**Comment by RudolphRiedel** [comment]
>    I didn't managed to group images, doesn't seems to work like this :
>
> EVE_cmd_setbitmap(0x800000 | (gauge0.addressFlash / 32), gauge0.format, gauge0.w, gauge0.h);
> EVE_cmd_dl(VERTEX2F(x0, y0));
> EVE_cmd_setbitmap(0x800000 | (gauge1.addressFlash / 32), gauge1.format, gauge1.w, gauge1.h);
> EVE_cmd_dl(VERTEX2F(x1, y1));
> EVE_cmd_dl(DL_END);
>
>Is it only for the _burst list building?

No, it works this way in non-burst mode as well, I am using it in my example.
The only thing missing from this block is
 EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);

>I tried to load the 6 images from flash and put them in the MEM_DL_STATIC but no chance again,
>could it be because part of the picture are out of screen bonds? (1024600 display, pictures are 304304)

Well, my MEM_DL_STATIC is defined like this:
#define MEM_DL_STATIC (EVE_RAM_G_SIZE - 4096)
This is the upper 4k of RAM_G.

Your MEM_GAUGES or whatever would be better off like this:
#define MEM_GAUGES 0x00000000

Well, or wherever else you like to place these at.

>I started looking at ESE, it's great tool to look at things fast! After playing a bit with it, I don't think the scrollbar is the feel I'm after,
> but I saw that there was a XY tracking coordinate. I'm going to try to make a Page+/ Page- by looking at the start XY touch
> and comparing it to the stop XY to get the direction

You could also check the REG_TOUCH_TAG_XY value for example if REG_TOUCH_TAG has a zero.
It reads as 32768 with no touch and has the current coordinates on touch.
Now store the first value and update a second value untill it goes back up to 32768.
The difference between the first and the last value shows a swipe to left or right.
Perhaps a threshold would be a good thing, like 100 pixels minimum.
